### PR TITLE
Add ansible-galaxy collection list command

### PR DESCRIPTION
**Pull request by samdoran**
##### SUMMARY

Add `ansible-galaxy collection list`.

##### ISSUE TYPE

- Feature Pull Request

##### COMPONENT NAME

`lib/ansible/cli/galaxy.py`

### CODE REVIEW

**Comment by samdoran** [comment]
bot_skip

File: lib/ansible/cli/galaxy.py [hunk 1]
```diff
@@ -969,39 +991,115 @@ def _display_role(gr):
...
+
+            # TODO: Should we support listing all collections in a given namespace?
+
+            # list a specific collection
+            if collection_name:
+                for path in collections_search_paths:
+                    collection_path = GalaxyCLI._resolve_path(path)
+                    if not os.path.exists(path):
+                        if path in C.COLLECTIONS_PATHS:
+                            # don't warn for missing default dirs
```
Comment (samdoran) [inline lib/ansible/cli/galaxy.py]: We don't do this for `roles`, meaning we show warnings when the default directories do not exist. I found that noisy, but maybe we should warn to be consistent? Or remove the warning when checking `roles`. I would like them at least be consistent.

File: lib/ansible/cli/galaxy.py [hunk 2]
```diff
@@ -969,39 +991,115 @@ def _display_role(gr):
...
+                        warnings.append("- the configured path {0} does not exist.".format(collection_path))
+                        continue
+                    elif not os.path.isdir(collection_path):
+                        warnings.append("- the configured path {0}, exists, but it is not a directory.".format(collection_path))
+                        continue
+
+                    path_found = True
+                    display.display("# {0}".format(path))
+
+                    # Make sure we look inside the 'ansible_collections' dir
```
Comment (samdoran) [inline lib/ansible/cli/galaxy.py]: This seems like a weird dance. Is there a better way to do this? Took me a while to figure out why I was having trouble getting at the actual collection directory.

File: lib/ansible/cli/galaxy.py [hunk 3]
```diff
@@ -969,39 +991,115 @@ def _display_role(gr):
...
+
+                    display.display("# {0}".format(path))
+
+                    # Make sure we look inside the 'ansible_collections' dir
+                    if os.path.split(path)[1] != 'ansible_collections':
+                        path = os.path.join(path, 'ansible_collections')
+
+                    collections = _find_existing_collections(path)
+                    # Sort collections by the namespace and name
+                    collections.sort(key=lambda x: '%s.%s' % (x.namespace, x.name))
```
Comment (samdoran) [inline lib/ansible/cli/galaxy.py]: The object's `repr` has the string I want, `namespace.collection`. Is there a way to use that and sort based on it rather than having to use a `lambda` here?

File: lib/ansible/cli/galaxy.py [hunk 4]
```diff
@@ -969,39 +991,111 @@ def _display_role(gr):
...
+
+                    try:
+                        namespace, collection = collection_name.split('.')
+                    except ValueError:
+                        raise AnsibleOptionsError("Incorrect value for collection name. Must be 'namespace.collection' got '%s'" % collection_name)
+
+                    b_collection_path = to_bytes(os.path.join(collection_path, namespace, collection), errors='surrogate_or_strict')
+
+                    if not os.path.exists(b_collection_path):
+                        warnings.append("- unable to find {0} in collection paths".format(collection_name))
```
Comment (s-hertel) [inline lib/ansible/cli/galaxy.py]: If there are multiple collection paths configured but a collection exists in only one of the paths, then `ansible-galaxy collection list namespace.collection` will warn for any paths that do not contain it. Should this warn only if the collection isn't found at all? Otherwise, maybe the collection path searched could be appended to this warning to make it more apparent where the collection was expected to be.

File: lib/ansible/cli/galaxy.py [hunk 5]
```diff
@@ -957,9 +975,13 @@ def execute_remove(self):
 
     def execute_list(self):
         """
-        lists the roles installed on the local system or matches a single role passed as an argument.
+        lists the roles or collections installed on the local system or matches a single role or collection passed as an argument.
         """
 
+        def _display_collection(collection):
+            fqcn = '%s.%s' % (collection.namespace, collection.name)
+            display.display('- {fqcn} ({version})'.format(fqcn=fqcn, version=collection.latest_version))
```
Comment (s-hertel) [inline lib/ansible/cli/galaxy.py]: I like this format (or `- namespace.collection_name:version`) a lot better than `- namespace.role_name, version` but I wonder if they should be consistent.

File: lib/ansible/cli/galaxy.py (hunk 1 above)
Comment (s-hertel) [reply lib/ansible/cli/galaxy.py]: I'd prefer less noise consistently too, but I don't have a strong opinion. You could probably move this `for path in collections_search_paths` + appending the warnings before `if collection_name:` so you don't need to write it for the `else` below too.

File: lib/ansible/cli/galaxy.py [hunk 6]
```diff
@@ -969,39 +991,111 @@ def _display_role(gr):
...
+
+                    display.display("# {0}".format(path))
+
+                    # Make sure we look inside the 'ansible_collections' dir
+                    if os.path.split(path)[1] != 'ansible_collections':
+                        path = os.path.join(path, 'ansible_collections')
+
+                    collections = _find_existing_collections(path)
+                    # Sort collections by the namespace and name
+                    collections.sort(key=lambda x: '%s.%s' % (x.namespace, x.name))
```
Comment (s-hertel) [inline lib/ansible/cli/galaxy.py]: You could use `collections.sort(key=to_text)`

File: lib/ansible/cli/galaxy.py [hunk 7]
```diff
@@ -969,39 +991,111 @@ def _display_role(gr):
...
+        warnings = []
+        galaxy_type = context.CLIARGS['type']
+        if context.CLIARGS['type'] == 'role':
+            if context.CLIARGS['role']:
+                # show the requested role, if it exists
+                name = context.CLIARGS['role']
+                gr = GalaxyRole(self.galaxy, self.api, name)
+                if gr.metadata:
+                    display.display('# %s' % os.path.dirname(gr.path))
+                    _display_role(gr)
```
Comment (s-hertel) [inline lib/ansible/cli/galaxy.py:1102]: path_found isn't set so this particular code path fails.

File: lib/ansible/cli/galaxy.py (hunk 2 above)
Comment (s-hertel) [reply lib/ansible/cli/galaxy.py]: I also had to do this in the verify PR. I'm not sure of a better way. `install` also does it. Maybe we should add a staticmethod below _resolve_path to also append ansible_collections (or possibly incorporate it in _resolve_path?) to reduce the duplicate code.

File: lib/ansible/cli/galaxy.py [hunk 8]
```diff
@@ -969,39 +991,111 @@ def _display_role(gr):
...
+            if collection_name:
+                for path in collections_search_paths:
+                    collection_path = GalaxyCLI._resolve_path(path)
+                    if not os.path.exists(path):
+                        if path in C.COLLECTIONS_PATHS:
+                            # don't warn for missing default dirs
+                            continue
+                        warnings.append("- the configured path {0} does not exist.".format(collection_path))
+                        continue
+                    elif not os.path.isdir(collection_path):
```
Comment (s-hertel) [inline lib/ansible/cli/galaxy.py]: Checking that the path exists and `isdir` should probably happen after possibly joining ansible_collections.

**Review by s-hertel — COMMENTED** [review]
This looks good. I had some minor, mostly cosmetic comments, but the overall functionality is quite nice.

File: lib/ansible/cli/galaxy.py (hunk 1 above)
Comment (samdoran) [reply lib/ansible/cli/galaxy.py]: I'm totally open to massaging the output format, but I agree they should be consistent.

File: lib/ansible/cli/galaxy.py (hunk 6 above)
Comment (samdoran) [reply lib/ansible/cli/galaxy.py]: That worked perfectly. Thanks!

File: lib/ansible/cli/galaxy.py [hunk 9]
```diff
@@ -22,8 +22,14 @@
 from ansible.errors import AnsibleError, AnsibleOptionsError
 from ansible.galaxy import Galaxy, get_collections_galaxy_meta_info
 from ansible.galaxy.api import GalaxyAPI
-from ansible.galaxy.collection import build_collection, install_collections, publish_collection, \
-    validate_collection_name
+from ansible.galaxy.collection import (
+    _find_existing_collections,
```
Comment (sivel) [inline lib/ansible/cli/galaxy.py]: Maybe this should be renamed to remove the `_`

File: lib/ansible/cli/galaxy.py [hunk 10]
```diff
@@ -957,9 +975,13 @@ def execute_remove(self):
 
     def execute_list(self):
         """
```
Comment (sivel) [inline lib/ansible/cli/galaxy.py:1078]: This PR increases the complexity of `execute_list` a lot:

devel:
```
galaxy.py:958:4: GalaxyCLI.execute_list (complexity 13)
```

This PR:
```
galaxy.py:976:4: GalaxyCLI.execute_list (complexity 31)
```

I'd recommend breaking this method up into a method for each "type". As it is, I feel like it makes it difficult to review.

**Review by sivel — CHANGES_REQUESTED** [review]

**Comment by samdoran** [comment]
The current output I have looks like this:
```
sdoran@mbp-2018 ~/S/ansible (galaxy-collection-list)> ansible-galaxy collection list
# /Users/sdoran/.ansible/collections/ansible_collections
Collection                                   Version
-------------------------------------------- -----------------
alikins.collection_inspect                   0.0.49
dynatrace_innovationlab.dynatrace_collection 1.0.6
mwhahaha.failure                             1.0.0
newswangerd.collection_demo                  1.0.10
pureport.pureport                            0.0.8
samdoran.hotpotato                           4.42.9.99.24.dev4
samdoran.icecubes                            2.7.15
samdoran.zztop                               3.14
sdoran@mbp-2018 ~/S/ansible (galaxy-collection-list)> ansible-galaxy role list
# /Users/sdoran/.ansible/roles
Role               Version
------------------ -------
geerlingguy.docker 2.6.0
samdoran.repo_epel 1.1
# /usr/local/share/ansible/roles
Role                Version
------------------- -------
ansible_pull        *
ansible_tower-bak   *
bootstrap_pi        *
caddy               *
certbot             *
deprole1            *
dnsmasq             *
docker              1.2.3
elasticsearch       *
fail2ban            *
firewall            *
fish                *
...
```

The column width is dynamic based on name and version length with a sane minimum.

I'm open to input or format or any other information that should be displayed for collections.

I do not know if changing the default output of `ansible-galaxy role list` is a good idea since there may be a lot of people parsing that output. I just wanted to at least have the option of making them unified, but that code may not ship with this change.

**Comment by samdoran** [comment]
@geerlingguy Do you have any opinions on the output of `collection list` and/or `role list`?

**Comment by geerlingguy** [comment]
@samdoran - I like it overall, though I think having one space before each of the paths (e.g. `# /Users/sdoran/.ansible/roles`) would be nice, because it's hard to visually parse the output with each path running onto the next. As someone who sometimes has 3 or more role paths configured, it could be messy if it's all one line after the other!

The only other value-add I can think of is—sometimes there is a role in more than one of the configured paths, it would be really cool if the output could show "this is the one that would be chosen" with an asterisk or something.

For collections, I haven't played around much with pathing, but isn't it possible you can have more than one path, too, and would have a similar conundrum?

File: lib/ansible/cli/galaxy.py (hunk 2 above)
Comment (samdoran) [reply lib/ansible/cli/galaxy.py]: Added a new function in #66441 to handle this so we can both use it.

File: lib/ansible/cli/galaxy.py (hunk 10 above)
Comment (samdoran) [reply lib/ansible/cli/galaxy.py:1078]: I agree, it's really bloated now. Now that I have it working, I'll simplify it.

File: lib/ansible/cli/galaxy.py (hunk 9 above)
Comment (samdoran) [reply lib/ansible/cli/galaxy.py]: Done in #66577

**Comment by samdoran** [comment]
@geerlingguy Do you mean:
```
 # /Users/sdoran/.ansible/roles
Role               Version
------------------ -------
geerlingguy.docker 2.6.0
samdoran.repo_epel 1.1
 # /usr/local/share/ansible/roles
Role                Version
------------------- -------
```

Or a newline between each path? This is how I have it currently:
```
> ansible-galaxy collection list

# /Users/sdoran/.ansible/collections/ansible_collections
Collection                                   Version
-------------------------------------------- -----------------
alikins.collection_inspect                   0.0.49
dynatrace_innovationlab.dynatrace_collection 1.0.6
newswangerd.collection_demo                  1.0.10
samdoran.hotpotato                           4.42.9.99.24.dev4
samdoran.icecubes                            2.7.15
samdoran.zztop                               3.14

# /usr/local/share/ansible/collections/ansible_collections
Collection        Version
----------------- -------
mwhahaha.failure  1.0.0
pureport.pureport 0.0.8
```

I think adding a leading space makes it look wonky:
```
> ansible-galaxy collection list

 # /Users/sdoran/.ansible/collections/ansible_collections
Collection                                   Version
-------------------------------------------- -----------------
alikins.collection_inspect                   0.0.49
dynatrace_innovationlab.dynatrace_collection 1.0.6
newswangerd.collection_demo                  1.0.10
samdoran.hotpotato                           4.42.9.99.24.dev4
samdoran.icecubes                            2.7.15
samdoran.zztop                               3.14

 # /usr/local/share/ansible/collections/ansible_collections
Collection        Version
----------------- -------
mwhahaha.failure  1.0.0
pureport.pureport 0.0.8
```

File: lib/ansible/cli/galaxy.py (hunk 4 above)
Comment (samdoran) [reply lib/ansible/cli/galaxy.py]: Ooo, good catch. I didn't have multiple collection paths in my test environment. I added that and now I see what you're talking about.

I think it should only warn if it's not found in _any_ of the paths and not warn for paths where it didn't exist.

Honestly, it may also make sense to warn if the same collection exists in _multiple_ paths, but I'm not going there for now.

**Comment by geerlingguy** [comment]
How you have it currently, then; no leading space, just a space (newline, that is, not technically a 'space') between each of the paths:

```
> ansible-galaxy collection list

# /Users/sdoran/.ansible/collections/ansible_collections
Collection                                   Version
-------------------------------------------- -----------------
alikins.collection_inspect                   0.0.49

# /usr/local/share/ansible/collections/ansible_collections
Collection        Version
----------------- -------
mwhahaha.failure  1.0.0
```

File: lib/ansible/cli/galaxy.py (hunk 5 above)
Comment (samdoran) [reply lib/ansible/cli/galaxy.py]: I would like them to be consistent. If anything, though, I would like to change the output of `role list` because it's not great.

I did the work of making the formats match, but I've moved the role output changes to a separate branch for now.

File: lib/ansible/cli/galaxy.py (hunk 4 above)
Comment (samdoran) [reply lib/ansible/cli/galaxy.py]: Fixed.

File: lib/ansible/cli/galaxy.py (hunk 7 above)
Comment (samdoran) [reply lib/ansible/cli/galaxy.py:1102]: > You could probably move this for path in collections_search_paths + appending the warnings before if collection_name: so you don't need to write it for the else below too.

Brilliant! I did that and the code is now far less redundant.

Comment (samdoran) [reply lib/ansible/cli/galaxy.py:1102]: > path_found isn't set so this particular code path fails.

I see what you mean. I need to rework the role section since the collection section is gathering everything and displaying at the end rather than as encountered.

Yet another reason I need to break these up. They are becoming far too intertwined.

File: lib/ansible/cli/galaxy.py (hunk 10 above)
Comment (samdoran) [reply lib/ansible/cli/galaxy.py:1078]: I have replaced `execute_list()` with two separate methods and moved the nested methods to private class methods. Things are looking much better now.
```
galaxy.py:1035:4: GalaxyCLI.execute_role_list (complexity 9)
galaxy.py:1077:4: GalaxyCLI.execute_collection_list (complexity 9)
```

File: lib/ansible/cli/galaxy.py [hunk 11]
```diff
@@ -166,11 +222,19 @@ def add_delete_options(self, parser, parents=None):
...
 
     def add_list_options(self, parser, parents=None):
+        galaxy_type = 'role'
+        if parser.metavar == 'COLLECTION_ACTION':
+            galaxy_type = 'collection'
+
+        # Seprate methods for listing a role or collection
+        list_func = getattr(self, 'execute_list_%s' % galaxy_type)
+
         list_parser = parser.add_parser('list', parents=parents,
```
Comment (s-hertel) [inline lib/ansible/cli/galaxy.py:231]: This would fix the sanity error.
```suggestion
         list_parser = parser.add_parser('list_%s' % galaxy_type, parents=parents,
```
I wonder if hacking/build_library/build_ansible/command_plugins/generate_man.py should inspect parser.get_default(dest='func') instead/additionally, but that's probably something to look into another time.

File: lib/ansible/cli/galaxy.py [hunk 12]
```diff
@@ -98,6 +146,13 @@ def init_parser(self):
                                 help='The path to the directory containing your roles. The default is the first '
                                      'writable one configured via DEFAULT_ROLES_PATH: %s ' % default_roles_path)
 
+        default_collections_path = C.config.get_configuration_definition('COLLECTIONS_PATHS').get('default', '')
+        collections_path = opt_help.argparse.ArgumentParser(add_help=False)
+        collections_path.add_argument('-p', '--collections-path', dest='collections_path', type=opt_help.unfrack_path(pathsep=True),
```
Comment (s-hertel) [inline lib/ansible/cli/galaxy.py]: Do you want to open a separate PR that does this since this is more common code we both need?

**Review by s-hertel — APPROVED** [review]
This looks really good. It seems like the integration tests are just failing due to the diff comparison with extra verbosity. Since the roles code has been refactored a bit there could be tests added to ensure backwards compatibility.

**Comment by AlanCoding** [comment]
I'm still digesting this, so I just wanted to leave some tentative comments right now.

----

The behavior for directions that do not contain roles vs. collections doesn't seem really consistent. No roles will give a warning.

```
$ ansible-galaxy role list
[WARNING]: - the configured path /Users/alancoding/.ansible/roles does not exist.
[WARNING]: - the configured path /usr/share/ansible/roles does not exist.
```

But collections will give this error

```
$ ansible-galaxy collection list -p nothing
ERROR! Unexpected Exception, this is probably a bug: [Errno 2] No such file or directory: b'/Users/alancoding/Documents/repos/ansible/nothing/ansible_collections'
to see the full traceback, use -vvv
```

This seems particularly problematic because I might do

```
$ mv ~/.ansible/collections/ansible_collections/ ~/.ansible/collections/ansible_collections_backup
$ ansible-galaxy collection list
ERROR! Unexpected Exception, this is probably a bug: [Errno 2] No such file or directory: b'/Users/alancoding/.ansible/collections/ansible_collections'
to see the full traceback, use -vvv
```

That doesn't prevent me from inspecting a folder, but it does prevent me from inspecting it without an error.

----

Here's a demo of the situation @geerlingguy was referencing:

```
$ ansible-galaxy collection list -p sam

# /Users/alancoding/Documents/repos/ansible/sam/ansible_collections
Collection                 Version
-------------------------- -------
alikins.collection_inspect 0.0.49
awx.awx                    9.0.1 
[WARNING]: Collection at '/Users/alancoding/.ansible/collections/ansible_collections/testns/testcoll' does not have a MANIFEST.json file, cannot detect version.

# /Users/alancoding/.ansible/collections/ansible_collections
Collection             Version
---------------------- -------
alancoding.awx_awesome 0.0.1
ansible.tower          3.6.0
awx.awx                9.1.0
testns.testcoll        * 
```

There are 2 different versions of the `awx.awx` collection, and when using this collections path, the older version will be used. I don't know if it would be more helpful to call out which _will_ be used, or call out which _will not_ be used, because the former is kind of a more notable and uncommon situation that requires user attention.

----

There's also not a ton of information here, just versions. It might be good to have checksums or something eventually, and an option for JSON output. Right now it seems pretty strictly geared for manual use.

**Comment by sivel** [comment]
> ERROR! Unexpected Exception, this is probably a bug: [Errno 2] No such file or directory: b'/Users/alancoding/.ansible/collections/ansible_collections'
> to see the full traceback, use -vvv

Tracebacks are never expected. I'd consider this a bug.

As for the 2 other features, I cannot guarantee we can add them right now.  For reference, the other new feature we are adding is `verify` and it will verify the collection that ansible will use, and give you the path.

**Comment by samdoran** [comment]
@AlanCoding Thank you for the feedback. Much appreciated!

> The behavior for directions that do not contain roles vs. collections doesn't seem really consistent. No roles will give a warning.

That's a bug. It shouldn't traceback.

>There are 2 different versions of the awx.awx collection, and when using this collections path, the older version will be used. I don't know if it would be more helpful to call out which will be used, or call out which will not be used, because the former is kind of a more notable and uncommon situation that requires user attention.

This would be rather tricky to do well since the listing logic has absolutely no relation to collection finding logic. At best, I'd have to try and mimmic what the Ansible engine is doing that in `collection list` in order to indicate which would be used. I have no idea how easy/hard that would be to do.

If you run `ansible-galaxy collection list specific.collection`, it will list _all_ instances of a collection in the search paths. So you at least get an indication there are multiple instances of a collection.

>There's also not a ton of information here, just versions.

I started with the same information we provide for `role list` (though I took the liberty of improving the output). I wasn't sure what other information would be helpful though we have discussed adding checksum/signature information. @s-hertel had the actual checksum displayed in the output of `validate` but we weren't sure how useful it was to actually display that. A user mostly cares "did the collection change" not "what's the exact checksum".

> It might be good to have checksums or something eventually, and an option for JSON output. Right now it seems pretty strictly geared for manual use.

The current output is meant for human consumption. For programmatic consumption, I would rather implement `--json`, which would alleviate a lot of the "What should I put in the output?" questions.

As @sivel mentioned, that's a stretch goal at this point. MVP is listing collection output for human consumption.


A side note on consistency between `role list` and `collection list`: I would like them to look and operate the same. I've elected to redesign the output formatting for `collection list` because there's no sense copying the current output of `role list` since it leaves a lot to be desired.

Once this PR is merged (and I have a good bit of work left to do), I have a separate branch that changes the `role list` output to match `collection list`. We can work on that later because that has its own list of issues to discuss.

File: lib/ansible/cli/galaxy.py (hunk 12 above)
Comment (samdoran) [reply lib/ansible/cli/galaxy.py]: Sure thing!

Comment (samdoran) [reply lib/ansible/cli/galaxy.py]: #66969

File: lib/ansible/cli/galaxy.py (hunk 11 above)
Comment (samdoran) [reply lib/ansible/cli/galaxy.py:231]: I was looking at fixing it in `hacking/build_library/build_ansible/command_plugins/generate_man.py` because maybe we want to generate man pages for the collection options too? I really don't know exactly what that file is doing, so your fix is probably the better option for now.

**Comment by samdoran** [comment]
> The behavior for directions that do not contain roles vs. collections doesn't seem really consistent. No roles will give a warning.

One more comment on this one: I decided not to warn if _default_ directories do not exist. `collection list` will only warn if a non-default path doesn't exist. I would like to change the `role list` behavior to not warn about missing default dirs, but that's for later.

**Comment by samdoran** [comment]
@AlanCoding I fixed the bug you reported and added a test for that scenario.

File: lib/ansible/cli/galaxy.py (hunk 11 above)
Comment (samdoran) [reply lib/ansible/cli/galaxy.py:231]: So this fixed the sanity test, but broke the CLI. I'm toying with some other ways to solve it.

Comment (samdoran) [reply lib/ansible/cli/galaxy.py:231]: Ok, I think I got it sorted. I put `execute_list()` back in and it just calls the correct underlying method.

**Comment by AlanCoding** [comment]
Confirmed the traceback mentioned earlier is resolved now.

The only other point I have is about help text. This command inspects multiple paths - really all of the paths in the config for collection paths. That's sensible, I don't disagree. But the help text doesn't really help.

```
  -p COLLECTIONS_PATH, --collections-path COLLECTIONS_PATH
                        The path to the directory containing your collections.
                        The default is the first writable one configured via
                        COLLECTIONS_PATHS:
                        ~/.ansible/collections:/usr/share/ansible/collections
```

Wording of "the first writable one" is erroneous for this command. It also further supports the incorrect assumption that the location given for `-p` is the only one inspected.

```
$ ansible-galaxy collection list -p sam awx.awx

# /Users/alancoding/Documents/repos/ansible/sam
Collection Version
---------- -------
awx.awx 9.0.1

# /Users/alancoding/.ansible/collections
Collection Version
---------- -------
awx.awx 9.1.0   
```

This inspects the folder sam, and the other locations in the config. There's a very non-obvious difference in the request `ansible-galaxy collection list -p sam` versus `ANSIBLE_COLLECTIONS_PATHS=sam ansible-galaxy collection list`.

**Comment by samdoran** [comment]
That's a very good point. The current behavior is it prepends the value passed to `-p` to the list of collections paths. The language "the first writeable one" is due to the option coming from common help rather than being `install` specific. That statement makes sense in the context of `install` but not `list` (or `validate`) for that matter.

So should the language be changed to more accurately describe what's happening, or should supplying `-p` behave the same as `ANSIBLE_COLLECTIONS_PATHS=sam ansible-galaxy collection list`?

Also, you found a bug in my display code with very short collection names. Thanks!

**Comment by samdoran** [comment]
@AlanCoding I have changed the help tex for `install -p` to be a more accurate description of what it does. Let me know if this more accurate:

```
  -p COLLECTIONS_PATH, --collection-path COLLECTIONS_PATH
                        Additional path to search for collections.
```

**Comment by AlanCoding** [comment]
Recent updates look perfect. I've pretty well exhausted my comments. 👍

**Review by sivel — APPROVED** [review]

File: docs/docsite/rst/galaxy/user_guide.rst [hunk 13]
```diff
@@ -78,6 +78,11 @@ Install multiple collections with a requirements file
 
 .. include:: ../shared_snippets/installing_multiple_collections.txt
 
+Listing Installed Collections
```
Comment (samccann) [inline docs/docsite/rst/galaxy/user_guide.rst]:
```suggestion
Listing installed collections
```

File: docs/docsite/rst/user_guide/collections_using.rst [hunk 14]
```diff
@@ -44,6 +44,80 @@ Configuring the ``ansible-galaxy`` client
 
 .. include:: ../shared_snippets/galaxy_server_list.txt
 
+.. _collections_listing:
+
+Listing Collections
```
Comment (samccann) [inline docs/docsite/rst/user_guide/collections_using.rst]:
```suggestion
Listing collections
```

File: docs/docsite/rst/user_guide/collections_using.rst [hunk 15]
```diff
@@ -44,6 +44,80 @@ Configuring the ``ansible-galaxy`` client
...
+      # /usr/share/ansible/collections/ansible_collections
+      Collection        Version
+      ----------------- -------
+      fortinet.fortios  1.0.6
+      pureport.pureport 0.0.8
+      sensu.sensu_go    1.3.0
+
+Run with ``-vvv`` to display more detailed information.
+
+To list a specific collection, pass a valid fully qualitifed collection name (FQCN) to the command ``ansible-galaxy collection list``. All instances of the collection will be listed.
```
Comment (samccann) [inline docs/docsite/rst/user_guide/collections_using.rst]:
```suggestion
To list a specific collection, pass a valid fully qualified collection name (FQCN) to the command ``ansible-galaxy collection list``. All instances of the collection will be listed.
```

File: docs/docsite/rst/user_guide/collections_using.rst [hunk 16]
```diff
@@ -44,6 +44,80 @@ Configuring the ``ansible-galaxy`` client
...
+      Collection       Version
+      ---------------- -------
+      fortinet.fortios 1.0.1
+
+      # /usr/share/ansible/collections/ansible_collections
+      Collection       Version
+      ---------------- -------
+      fortinet.fortios 1.0.6
+
+To search other paths for collections, use the ``-p`` option. Specify multiple search paths by separating them with a ``:``. The list of paths specified on the command line will be added to the begining of the configured collections search paths.
```
Comment (samccann) [inline docs/docsite/rst/user_guide/collections_using.rst]:
```suggestion
To search other paths for collections, use the ``-p`` option. Specify multiple search paths by separating them with a ``:``. The list of paths specified on the command line will be added to the beginning of the configured collections search paths.
```

**Review by samccann — COMMENTED** [review]
just a few docs nits

**Comment by sivel** [comment]
HAPPY MERGE DAY!
